MIT_6.S081/2020/COMMON_ERRORS.md

```diff
@@ -0,0 +1,143 @@
+# Common Errors
+
+After installing all the required deps and you run `make qemu`, you will face this error:
+```bash
+user/sh.c: In function 'runcmd':
+user/sh.c:60:1: error: infinite recursion detected [-Werror=infinite-recursion]
+   60 | runcmd(struct cmd *cmd)
+      | ^~~~~~
+user/sh.c:91:5: note: recursive call
+   91 |     runcmd(rcmd->cmd);
+      |     ^~~~~~~~~~~~~~~~~
+user/sh.c:111:7: note: recursive call
+  111 |       runcmd(pcmd->left);
+      |       ^~~~~~~~~~~~~~~~~~
+user/sh.c:118:7: note: recursive call
+  118 |       runcmd(pcmd->right);
+      |       ^~~~~~~~~~~~~~~~~~~
+user/sh.c:97:7: note: recursive call
+   97 |       runcmd(lcmd->left);
+      |       ^~~~~~~~~~~~~~~~~~
+user/sh.c:99:5: note: recursive call
+   99 |     runcmd(lcmd->right);
+      |     ^~~~~~~~~~~~~~~~~~~
+user/sh.c:129:7: note: recursive call
+  129 |       runcmd(bcmd->cmd);
+      |       ^~~~~~~~~~~~~~~~~
+cc1: all warnings being treated as errors
+make: *** [user/sh.o] Error 1
+```
+
+This is because in the `Makefile` `CFLAGS`, we have the `-Werror` flag (line 70) which tells the compiler (gcc/clang) to treat all warnings as errors.
+This means any warnigs that would normally just display a message will instead cause the compilation to fail.
+
+## Fix
+
+1. Go to the file `user/sh.c` and find the function `runcmd(struct cmd *cmd)`
+
+2. Make the following file changes to it:
+
+```C
```

Comment:
For better clarity, can use the diff instead of the whole file?

```diff
diff --git a/user/sh.c b/user/sh.c
index 83dd513..b8fa790 100644
--- a/user/sh.c
+++ b/user/sh.c
@@ -54,6 +54,8 @@ void panic(char*);
 struct cmd *parsecmd(char*);

 // Execute cmd.  Never returns.
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Winfinite-recursion"
 void
 runcmd(struct cmd *cmd)
 {
 ```

Reply:
done